pkg/provisioner/ironic/ironic.go

```diff
@@ -1129,6 +1129,71 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	if p.bmcAccess.RAIDInterface() != "" {
+		// Set raid configuration
+		err = setTargetRAIDCfg(p, ironicNode)
+		if err != nil {
+			p.log.Error(err, "failed in setting the target RAID configuration")
+			return
+		}
+		// Build raid clean steps
+		cleanSteps = append(cleanSteps, buildRAIDCleanSteps(p.host.Status.Provisioning.RAID)...)
+	} else if p.host.Status.Provisioning.RAID != nil {
+		return result, fmt.Errorf("the BMC(%v) of the current node does not support RAID", p.bmcAccess.Driver())
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		result.Dirty = !success
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.CleanWait, nodes.Cleaning, nodes.Available, nodes.Active:
+		// Do Nothing (go to provisioning state)
```

Comment:
Wouldn't be better to wait for the successful completion of the (manual) cleaning before moving to the Provisioning state? This will make also the failure management a little bit easier, I guess

Reply:
Yes, but according to the state machine diagram of ironic:

![states.svg](https://docs.openstack.org/ironic/latest/_images/states.svg)

The state before and after **cleaning** are both **mangeable**, this is difficult to handle in a single state.

Reply:
Right, but if feasible I'd stick to Ironic to determine the current situation. From the [documentation on manual cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#manual-cleaning) it looks like that `driver_internal_info` could be inspected to determine which steps is currently performing (or failed), maybe it could be used to determine if the manual cleaning has been completed.
@dtantsur do you know if there's a better way to determine that the manual cleaning completed successfully?